### PR TITLE
Documentation: Document which file formats exist

### PR DESCRIPTION
**Pull request by BenWiederhake**
I intentionally don't document the format specifics, to prevent this
file from becoming outdated. Hence, only a list with pointers to where
to find details.

### CODE REVIEW

File: Documentation/FileFormats.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+`width`, and `height` (encoded as a decimal number).
+
+The data is encoded in width times height many bytes, either 0 (clear) or 1 (set).
+
+## image/x-serenityos (Clipboard-only)
+
+Requires the metadata-fields `width`, `height`, `scale`, `format` (see `Gfx::BitmapFormat`), and `pitch`, encoded as decimal strings.
+
+The data is encoded according to `Gfx::determine_storage_format(BitmapFormat)`, so either as
+BGRx8888, BGRA8888, RGBA8888, or 8-bit pallette index. Note that the pallette is not transferred.
```
Comment (linusg) [inline Documentation/FileFormats.md:17]:
```suggestion
BGRx8888, BGRA8888, RGBA8888, or 8-bit palette index. Note that the palette is not transferred.
```

(https://www.vocabulary.com/dictionary/pallette :^))

File: Documentation/FileFormats.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Serenity Formats
+
+This document summarizes all non-standard file formats, data formats, and mime-types used by Serenity.
+
+## glyph/x-fonteditor (Clipboard-only)
+
+Requires the metadata-fields, `char` (usually encoded as a string of length 1: the codepoint as a character),
+`width`, and `height` (encoded as a decimal number).
+
+The data is encoded in width times height many bytes, either 0 (clear) or 1 (set).
```
Comment (linusg) [inline Documentation/FileFormats.md:10]: This is a bit vague, and I'd have to look up how it works. Is it possible to give an actual example in textual representation? How are these fields stored?

File: Documentation/FileFormats.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+## image/x-serenityos (Clipboard-only)
+
+Requires the metadata-fields `width`, `height`, `scale`, `format` (see `Gfx::BitmapFormat`), and `pitch`, encoded as decimal strings.
+
+The data is encoded according to `Gfx::determine_storage_format(BitmapFormat)`, so either as
+BGRx8888, BGRA8888, RGBA8888, or 8-bit pallette index. Note that the pallette is not transferred.
+
+## text/uri-list (Clipboard-only)
+
+Newline-delimited set of URIs.
```
Comment (linusg) [inline Documentation/FileFormats.md:21]: would also appreciate an example here, e.g.

```suggestion
Newline-delimited set of URIs.

Example:

```
http://serenityos.org/
https://example.com/foo/bar
```
```

(right?)

File: Documentation/FileFormats.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+## Application File (`*.af` files)
+
+These files are human-readable and are a subset of the INI-format, have no easily detectable filemagic, and define System Menu entries.
```
Comment (linusg) [inline Documentation/FileFormats.md:25]:
```suggestion
These files are human-readable and are a subset of the INI-format, have no easily detectable filemagic, and define System Menu entries.

They are stored in [`/res/apps`](../Base/res/apps).
```

File: Documentation/FileFormats.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+## Font (`*.font` files)
+
+These files contain bitmap definitions of fonts, either varying-width or fixed-width.
+The header definition can be found in `Gfx::FontFileHeader` in `Userland/Libraries/LibGfx/BitmapFont.cpp`.
```
Comment (linusg) [inline Documentation/FileFormats.md:30]: I always like linking files for easy access:

```suggestion
The header definition can be found in `Gfx::FontFileHeader` in [`Userland/Libraries/LibGfx/BitmapFont.cpp`](../Userland/Libraries/LibGfx/BitmapFont.cpp).
```

File: Documentation/FileFormats.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+Most prominently, the first four bytes contain the filemagic: `!Fnt`.
+
+## GUI Markup Language (`*.gml` files)
+
+These files are human-readable, have no easily detectable filemagic, and define GUI interfaces.
+The format is strongly influenced by QML, the Qt Modeling Lange.
```
Comment (linusg) [inline Documentation/FileFormats.md:36]:
```suggestion
The format is strongly influenced by QML, the Qt Modeling Language.
```

File: Documentation/FileFormats.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+## Inter Process Communication (`*.ipc` files)
+
+These files are human-readable, have no easily detectable filemagic, and define IPC interfaces.
+The format is loosely inspired by C++ headers.
+
+## Inter Process Communication (through Unix sockets)
+
+The various services each have their own formats, all very similar, and automatically implemented through LibIPC. The specifics depend on the corresponding source `.ipc` file.
+
+The format can be identified by the format magic, which is derived in `Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp`
```
Comment (linusg) [inline Documentation/FileFormats.md:47]:
```suggestion
The format can be identified by the format magic, which is derived in [`Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp`](../Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp)
```

**Review by linusg — COMMENTED** [review]
Great initiative!

File: Documentation/FileFormats.md (hunk 2 above)
Comment (BenWiederhake) [reply Documentation/FileFormats.md]: Can you specify which part is vague? Or is it the overall thing? Maybe it helps to phrase it like this?: "The glyph bitmap data is encoded …"

In general, I try to not be too specific about how it works; this document should just be an entry point to learn about which formats exist, and where to learn the details.

File: Documentation/FileFormats.md (hunk 3 above)
Comment (BenWiederhake) [reply Documentation/FileFormats.md:25]: No, but I'll replace it with a better example, and a list of applications that support it.

File: Documentation/FileFormats.md (hunk 4 above)
Comment (BenWiederhake) [reply Documentation/FileFormats.md:29]: Eh, alright, I'll add it in, but I would normally assume the user is able to run `git ls-files *.af` by themselves.

File: Documentation/FileFormats.md (hunk 2 above)
Comment (linusg) [reply Documentation/FileFormats.md]: > Can you specify which part is vague?

Whether the metadata is part of the payload or not. I did look at the source, and it's not - just a regular `HashMap` passed to `GUI::Clipboard::the().set_data()`. With that as context, this makes much more sense, so maybe:

> Requires the following fields to be present in the clipboard metadata `HashMap`: ...

?

File: Documentation/FileFormats.md (hunk 6 above)
Comment (BenWiederhake) [reply Documentation/FileFormats.md:40]: It's typos like these that make me quest my sanity ;)

File: Documentation/FileFormats.md (hunk 4 above)
Comment (linusg) [reply Documentation/FileFormats.md:29]: Sure, everyone can do that. But you forget that this file is also viewable online, and hyperlinks are a nice thing :^)
I'd prefer a simple link over navigating there using the GitHub UI.

File: Documentation/FileFormats.md (hunk 3 above)
Comment (linusg) [reply Documentation/FileFormats.md:25]: oh, URIs, not URLs. Yeah, I remember what this is about now. Proper example would be nice, thanks :)

**Comment by BenWiederhake** [comment]
Changes since initial push:
- Added nice "see also"s with pointers to the right files/directories
- Added examples and fixed typos as suggested by linusg (thanks!)
- Rebased

File: Documentation/FileFormats.md (hunk 2 above)
Comment (BenWiederhake) [reply Documentation/FileFormats.md]: Right, in this document I should not assume that the reader already knows how the clipboard works.

**Comment by BenWiederhake** [comment]
Changes since last push:
- Documented rough outline of how clipboard and drag & drop work, so that the clipboard "metadata" makes more sense.
- Rebased

**Pull request merged**